### PR TITLE
sqlite: fix cancellation cleanup for ExecContext

### PR DESCRIPTION
**Pull request by creachadair**
I did not test the changes in #103 sufficiently, and missed a corner case in
ExecContext: We need to make sure the context cleanup happens prior to the
synchronization, not in a defer.

Add a test to actually trigger this case, and fix the cleanup so it happens
before return as intended.

### CODE REVIEW

File: sqlite.go [hunk 1]
```diff
@@ -490,23 +490,22 @@ func (s *stmt) ExecContext(ctx context.Context, args []driver.NamedValue) (drive
...
 			defer close(done)
 			if ctx.Err() != nil {
 				db.Interrupt()
 			}
 		})
-		// In the event we get an error from the query's initial execution of
-		// sqlite3_step below and exit early, dissociate the cancellation since
-		// we don't want it to fire and potentially stop a later execution.
-		defer stop()
+		wait = func() { pcancel(); <-done }
```
Comment (maisem) [inline sqlite.go:508]:
```
defer func() {
  pcancel()
  <-done
}()
```

would also work

File: sqlite.go [hunk 2]
```diff
@@ -490,23 +490,22 @@ func (s *stmt) ExecContext(ctx context.Context, args []driver.NamedValue) (drive
...
-		done = make(chan struct{})
+		done := make(chan struct{})
 		pctx, pcancel := context.WithCancel(ctx)
-		defer pcancel() // to make the AfterFunc fire and close(done)
-
 		db := s.stmt.DBHandle()
-		stop := context.AfterFunc(pctx, func() {
+		context.AfterFunc(pctx, func() {
 			defer close(done)
 			if ctx.Err() != nil {
```
Comment (maisem) [inline sqlite.go:504]: write a comment here why `ctx` and not `pctx`?

File: sqlite.go (hunk 1 above)
Comment (maisem) [reply sqlite.go:508]: or if you prefer
```
wait := func() { pcancel(); <-done }
defer wait()
```

Comment (maisem) [reply sqlite.go:508]: fine as is too, just another approach

File: sqlite.go (hunk 2 above)
Comment (creachadair) [reply sqlite.go:507]: Good idea, done (also in the other spot we do this).

**Review by maisem — APPROVED** [review]
Thank you!

File: sqlite.go (hunk 1 above)
Comment (creachadair) [reply sqlite.go:511]: It won't quite work the way you describe it—if we bind the cleanup function in this scope, we can't call it below, and it's essential we do so prior to returning (since, as you noted in Slack, the defer doesn't happen till too late)

Comment (creachadair) [reply sqlite.go]: Per our offline conversation: Actually in this case a defer is fine, because the code to check the results doesn't depend on the cancellation being resolved. So, I swapped it to a defer (since that avoids having to think about scopes at all). Thanks!

**Review by maisem — APPROVED** [review]

**Review by bradfitz — COMMENTED** [review]
Nice. And more readable now too as a bonus.
